Extensions/ArtifactEngine/Engine/artifactEngine.ts

```diff
@@ -8,37 +11,61 @@ import { ArtifactItemStore } from '../Store/artifactItemStore';
+            artifactEngineOptions.artifactCacheDirectory = artifactEngineOptions.artifactCacheDirectory ? artifactEngineOptions.artifactCacheDirectory : tl.getVariable("AGENT_WORKFOLDER");
             this.createPatternList(artifactEngineOptions);
+            var artifactName = sourceProvider.getRootItemPath();
+            this.artifactItemStore = new ArtifactItemStore();
             this.artifactItemStore.flush();
             Logger.verbose = artifactEngineOptions.verbose;
             this.logger = new Logger(this.artifactItemStore);
             this.logger.logProgress();
             sourceProvider.artifactItemStore = this.artifactItemStore;
             destProvider.artifactItemStore = this.artifactItemStore;
+            this.cacheProvider = new CacheProvider(artifactEngineOptions.artifactCacheDirectory,artifactEngineOptions.artifactCacheKey,artifactName);
             sourceProvider.getRootItems().then((itemsToProcess: models.ArtifactItem[]) => {
                 this.artifactItemStore.addItems(itemsToProcess);
-
-                for (let i = 0; i < artifactEngineOptions.parallelProcessingLimit; ++i) {
-                    var worker = new Worker<models.ArtifactItem>(i + 1, item => this.processArtifactItem(sourceProvider, item, destProvider, artifactEngineOptions), () => this.artifactItemStore.getNextItemToProcess(), () => !this.artifactItemStore.itemsPendingProcessing());
-                    workers.push(worker.init());
-                }
-
-                Promise.all(workers).then(() => {
-                    this.logger.logSummary();
-                    sourceProvider.dispose();
-                    destProvider.dispose();
-                    resolve(this.artifactItemStore.getTickets());
-                }, (err) => {
-                    ci.publishEvent('reliability', <ci.IReliabilityData>{ issueType: 'error', errorMessage: JSON.stringify(err, Object.getOwnPropertyNames(err)) });
-                    sourceProvider.dispose();
-                    destProvider.dispose();
-                    reject(err);
+                this.createNewHashMap(sourceProvider,itemsToProcess).then(() => {
+                    for (let i = 0; i < artifactEngineOptions.parallelProcessingLimit; ++i) {
+                        var worker = new Worker<models.ArtifactItem>(i + 1, item => this.processArtifactItem(sourceProvider, item, destProvider, artifactEngineOptions), () => this.artifactItemStore.getNextItemToProcess(), () => !this.artifactItemStore.itemsPendingProcessing());
+                        workers.push(worker.init());
+                    }
+        
+                    Promise.all(workers).then(() => {
+                        this.logger.logSummary();
+                        
+                        var destination = destProvider.getRootLocation();
+                        var artifactName = sourceProvider.getRootItemPath();                        
```

Comment:
can we move hash validations to artifactitemstore, basically destination provider will generate the hash along with writing the item and return back the hash as part of item, while updating the state of item in store it can check that the actual and downloaded hashes match. This way we will be able to retry in case hash validation fails also hash validations will not be limited to destination provider being filesystem (though we will add the support only for filesystem currently).

Talk to gyan if you have doubts regarding this approach or give me a call.